src/System.Linq.Expressions/src/System/Linq/Expressions/BlockExpression.cs

```diff
@@ -937,5 +955,20 @@ internal static void ValidateVariables(ReadOnlyCollection<ParameterExpression> v
                 set.Add(v);
             }
         }
+
+        private static BlockExpression GetOptimizedBlockExpression(IReadOnlyList<Expression> expressions)
+        {
+            switch (expressions.Count)
+            {
+                case 2: return Block(expressions[0], expressions[1]);
+                case 3: return Block(expressions[0], expressions[1], expressions[2]);
+                case 4: return Block(expressions[0], expressions[1], expressions[2], expressions[3]);
+                case 5: return Block(expressions[0], expressions[1], expressions[2], expressions[3], expressions[4]);
+                default:
+                    ContractUtils.RequiresNotEmptyList(expressions, "expressions");
+                    RequiresCanRead(expressions, "expressions");
+                    return new BlockN(expressions.ToArray());
```

Comment:
Actually, no, it makes sense to create node with an expression array. 
The array will be automatically wrapped into ReadOnly, but only if GetOrMakeExpressions is called, which is not considered common. Otherwise expressions will be stored as array, which is leaner